src/app/tests/TestMessageDef.cpp

```diff
@@ -0,0 +1,1125 @@
+
+    chip::System::PacketBuffer::Free(buf);
+}
+
+void CommandDataElementTest(nlTestSuite * apSuite, void * apContext)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    CommandDataElement::Builder commandDataElementBuilder;
+    CommandDataElement::Parser commandDataElementParser;
+    chip::TLV::TLVWriter writer;
+    chip::TLV::TLVReader reader;
+    chip::System::PacketBuffer * buf = chip::System::PacketBuffer::New();
+
+    writer.Init(buf);
+    commandDataElementBuilder.Init(&writer);
+    BuildCommandDataElement(apSuite, commandDataElementBuilder);
+    err = writer.Finalize();
+    NL_TEST_ASSERT(apSuite, err == CHIP_NO_ERROR);
+
+    DebugPrettyPrint(buf);
+
+    reader.Init(buf);
+    err = reader.Next();
+    NL_TEST_ASSERT(apSuite, err == CHIP_NO_ERROR);
+
+    commandDataElementParser.Init(reader);
+    ParseCommandDataElement(apSuite, commandDataElementParser);
+
+    chip::System::PacketBuffer::Free(buf);
+}
+
+void CommandListTest(nlTestSuite * apSuite, void * apContext)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    chip::TLV::TLVWriter writer;
+    chip::TLV::TLVReader reader;
+    CommandList::Builder commandListBuilder;
+    chip::System::PacketBuffer * buf = chip::System::PacketBuffer::New();
+
+    writer.Init(buf);
+    commandListBuilder.Init(&writer);
+    BuildCommandList(apSuite, commandListBuilder);
+    err = writer.Finalize();
+    NL_TEST_ASSERT(apSuite, err == CHIP_NO_ERROR);
+
+    DebugPrettyPrint(buf);
+
+    reader.Init(buf);
+    err = reader.Next();
+    NL_TEST_ASSERT(apSuite, err == CHIP_NO_ERROR);
+    ParseCommandList(apSuite, reader);
+
+    chip::System::PacketBuffer::Free(buf);
+}
+
+void ReportDataTest(nlTestSuite * apSuite, void * apContext)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    chip::TLV::TLVWriter writer;
+    chip::TLV::TLVReader reader;
+    chip::System::PacketBuffer * buf = chip::System::PacketBuffer::New();
+
+    writer.Init(buf);
+    BuildReportData(apSuite, writer);
+    err = writer.Finalize();
+    NL_TEST_ASSERT(apSuite, err == CHIP_NO_ERROR);
+
+    DebugPrettyPrint(buf);
+
+    reader.Init(buf);
+    err = reader.Next();
+    NL_TEST_ASSERT(apSuite, err == CHIP_NO_ERROR);
+    ParseReportData(apSuite, reader);
+
+    chip::System::PacketBuffer::Free(buf);
+}
+
+void InvokeCommandTest(nlTestSuite * apSuite, void * apContext)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    chip::TLV::TLVWriter writer;
+    chip::TLV::TLVReader reader;
+    chip::System::PacketBuffer * buf = chip::System::PacketBuffer::New();
+
+    writer.Init(buf);
+    BuildInvokeCommand(apSuite, writer);
+    err = writer.Finalize();
+    NL_TEST_ASSERT(apSuite, err == CHIP_NO_ERROR);
+
+    DebugPrettyPrint(buf);
+
+    reader.Init(buf);
+    err = reader.Next();
+    NL_TEST_ASSERT(apSuite, err == CHIP_NO_ERROR);
+    ParseInvokeCommand(apSuite, reader);
+
+    chip::System::PacketBuffer::Free(buf);
+}
+
+/**
+ *   Test Suite. It lists all the test functions.
+ */
+
+// clang-format off
+const nlTest sTests[] =
```

Comment:
I don't see any adversarial tests or verification that error cases return reasonable errors.  Are more than functional tests coming at some point?

Reply:
Yep, I will add more adversarial tests in the follow-up ticket, https://github.com/project-chip/connectedhomeip/issues/3886  Currently I am only adding functional test in this PR. thanks